release: 0.3.0

DifferentialOrange opened the pull request:
## Overview

This release introduces new API to check module version in code.

## Changed
* Add versioning support through module._VERSION (PR #65).
* Deprecate versioning support through module.VERSION (PR #66).

Totktonada commented:
@Khatskevich @SudoBobo The PR is ready for review.

Khatskevich commented on test/testdata/nullable_1_1_conn_testdata.lua:
```diff
@@ -0,0 +1,398 @@
+    end
+
+    results = results .. print_and_return(([[
+        +---------------------+
```
What is that?) Why do you print it? Moreover it is not printed in shard tests. Better to just leave it as a comment?

Totktonada replied:
Forgot to print in shard tests. Will do. It was for convenient investigation of run of the test.

Totktonada replied:
Consider yaml module in tarantool:

third_party/lua-yaml/lyaml.cc
```
         } else if (strstr(str, "\n\n") != NULL) {
            /*  
             * Tarantool-specific: use literal style for string with empty lines.
             * Useful for tutorial().
             */
            style = YAML_LITERAL_SCALAR_STYLE;
         }   
```

I added two newlines in the begin of the string and got right output for the shard tests.

Khatskevich commented on test/testdata/nullable_1_1_conn_testdata.lua:
```diff
@@ -0,0 +1,398 @@
+-- The example was inspired by [1]. Consider [2] for the problem description
```
It is up to you, but I do not like those teste. In my opinion we need to:
1. Write common tests
2. Check results in tests

Pros:
1. To validate test results we do not need to read 3 result files
2. The code you copy can contain small bugs, so, it should also be read 3 times
It is just a waste of time!

Totktonada replied:
I think I should work on reducing code size of the tests. Maybe I should consider 'common' test suite + tap13, but I want to: run local test as just file, don’t repeat the sime initialization / running steps, working shard module reloading. I’ll consider your pain at reading in the next PRs.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -307,8 +307,8 @@ gql_type = function(state, avro_schema, collection, collection_name)
-            assert(c.type == '1:1' or c.type == '1:N',
-                'connection.type must be 1:1 or 1:N, got ' .. c.type)
+            assert(c.type == '1:1' or c.type == '1:1*' or c.type == '1:N',
+                'connection.type must be 1:1, 1:1* or 1:N, got ' .. c.type)
```
I want to be consistent with our extended avro-schema approach.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -253,7 +253,7 @@ end
 ---
 --- @tparam table state for read state.accessor and previously filled
---- state.types (state.types are gql types)
+--- state.nullable_collection_types (its are gql types)
```
Fixed.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -230,7 +230,7 @@ end
 ---
 --- @tparam table state for read state.accessor and previously filled
---- state.types
+--- state.nullable_collection_types
```
I generally against common words to refer specific things. This field contains only collection types (and does not contain, say, nested records) and that types converted to nullable ones (because of lazy types evaluation, consider comments). So more specific name `nullable_collection_types` is fair term for that. Maybe `collection_types_nullable` is better, though.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -319,16 +319,21 @@ gql_type = function(state, avro_schema, collection, collection_name)
             local c_args
             if c.type == '1:1' then
+                destination_type = types.nonNull(
+                    state.nullable_collection_types[c.destination_collection])
```
1. Fixed.
2. I’m sure that collection types are nullable because it is how I save it.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -354,8 +360,43 @@ gql_type = function(state, avro_schema, collection, collection_name)
+                    local is_full_match_ok = are_all_parts_null or
+                        are_all_parts_non_null
+                    if not is_full_match_ok then -- avoid extra json.encode()
+                        assert(is_full_match_ok,
```
Please, read the comment for this if.

Totktonada commented on test/testdata/nullable_1_1_conn_testdata.lua:
```diff
@@ -0,0 +1,400 @@
+    end)
+
+    -- FULL MATCH constraint: connection key parts must be all non-nulls or all
+    -- nulls
```
I have two failing tests here. Commented explicitly that they are fails.

Totktonada commented on test/testdata/nullable_1_1_conn_testdata.lua:
```diff
@@ -0,0 +1,400 @@
+
+    local function new_email(body)
+        return {
+            localpart = random_string(16):hex(),
```
I initialized PRNG with a constant seed. You can use it in the output and it will be the same.

Totktonada replied:
Found that math.random() gives different numbers on different machines. Implemented simple PRNG in the test.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -354,8 +360,43 @@ gql_type = function(state, avro_schema, collection, collection_name)
+                    -- key parts
+                    --
+                    -- zero parts count considered as ok by this check
+                    local is_full_match_ok = are_all_parts_null or
```
Ok, I’ll use name `ok`. Don’t want to carry lines, it looks ugly here.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -354,8 +360,43 @@ gql_type = function(state, avro_schema, collection, collection_name)
+                            value
+
+                        if value ~= nil then -- nil or box.NULL
+                            are_all_parts_null = false
```
I’ll think about that before merge, but I’ll skip it now. Thanks for the point.

Totktonada replied:
It seems that similar effort was made by @SudoBobo in the scope of #8. I’m going to leave it as is for now. Feel free to file an issue re code refactoring if you want to track this.

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -354,8 +360,43 @@ gql_type = function(state, avro_schema, collection, collection_name)
+                    --   return the empty list.
+                    if are_all_parts_null then
+                        assert(c.type == '1:1*' or c.type == '1:N',
+                            'only 1:1* or 1:N connections can have ' ..
```
We discussed the line carrying in the case in context of tarantool code style and decided to force formal line lenght limit. So I’ll prefer to fit the limit, but I’ll try to split a message by meaningful parts.

extra info: TBD

Totktonada replied:
extra info: added

Totktonada commented on graphql/tarantool_graphql.lua:
```diff
@@ -354,8 +360,43 @@ gql_type = function(state, avro_schema, collection, collection_name)
+                    end
+
+                    -- check FULL match constraint before request of
+                    -- destination object(s)
```
My approach is the following. I use small letter in the begin and omit period at the end when the comment is one clause or like so (maybe even an expanded sentence, but only one). Several sentences started from capital letter and have period at the end.

Here we have several sentences, so I reformatted this into one multi-sentence comment (capital letter + period).

DifferentialOrange committed:
release: 0.3.0

Overview

  This release introduces new API to check module version in code.

Changed
  * Add versioning support through module._VERSION (PR #65).
  * Deprecate versioning support through module.VERSION (PR #66).
